dbms/src/Storages/DeltaMerge/VersionChain/DMFileHandleIndex.h

```diff
@@ -0,0 +1,241 @@
+        }
+        clipped_need_read_packs = std::move(calc_read_packs);
+    }
+
+    // Release handle column data to save memory.
+    void cleanHandleColumn()
+    {
+        std::fill(clipped_handle_packs.begin(), clipped_handle_packs.end(), nullptr);
+        clipped_need_read_packs = std::vector<UInt8>(clipped_pack_range.count(), 1);
+    }
+
+private:
+    // Find the pack may contian the given handle.
+    std::optional<UInt32> getClippedPackId(HandleRefType h) const
+    {
+        if (unlikely(rowkey_range && !inRowKeyRange(*rowkey_range, h)))
+            return {};
+
+        auto itr = std::lower_bound(clipped_pack_index.begin(), clipped_pack_index.end(), h);
+        if (itr == clipped_pack_index.end())
+            return {};
+        return itr - clipped_pack_index.begin();
+    }
+
+    // Find the base(oldest) version of the given handle in the given pack.
+    std::optional<RowID> getBaseVersion(const DMContext & dm_context, HandleRefType h, UInt32 clipped_pack_id)
+    {
+        loadHandleIfNotLoaded(dm_context);
+        ColumnView<HandleType> handle_col(*clipped_handle_packs[clipped_pack_id]);
+        auto itr = std::lower_bound(handle_col.begin(), handle_col.end(), h);
+        if (itr != handle_col.end() && *itr == h)
+            return itr - handle_col.begin() + clipped_pack_offsets[clipped_pack_id];
+        return {};
+    }
+
+    std::vector<HandleType> loadPackIndex(const DMContext & dm_context) const
+    {
+        auto max_values = loadPackMaxValue<HandleType>(dm_context, *dmfile, MutSup::extra_handle_id);
+        return std::vector<HandleType>(
+            max_values.begin() + clipped_pack_range.start_pack_id,
+            max_values.begin() + clipped_pack_range.end_pack_id);
+    }
+
+    std::vector<UInt32> loadPackOffsets() const
+    {
+        const auto & pack_stats = dmfile->getPackStats();
+        std::vector<UInt32> pack_offsets(clipped_pack_range.count(), 0);
+        for (UInt32 clipped_pack_id = 1; clipped_pack_id < pack_offsets.size(); ++clipped_pack_id)
+        {
+            UInt32 real_pack_id = clipped_pack_id + clipped_pack_range.start_pack_id;
+            pack_offsets[clipped_pack_id] = pack_offsets[clipped_pack_id - 1] + pack_stats[real_pack_id - 1].rows;
+        }
+        return pack_offsets;
+    }
+
+    void loadHandleIfNotLoaded(const DMContext & dm_context)
+    {
+        if (likely(clipped_need_read_packs.empty()))
+            return;
+
+        auto read_pack_ids = std::make_shared<IdSet>();
+        for (UInt32 i = 0; i < clipped_need_read_packs.size(); ++i)
+        {
+            if (clipped_need_read_packs[i])
+                read_pack_ids->insert(i + clipped_pack_range.start_pack_id);
+        }
+
+        auto pack_filter = DMFilePackFilter::loadFrom(
+            dm_context,
+            dmfile,
+            /*set_cache_if_miss*/ true,
+            /*rowkey_ranges*/ {}, // Empty means all
+            EMPTY_RS_OPERATOR,
+            read_pack_ids);
```

Comment:
Seems we need to load the index multiple times, in `loadPackIndex` and `getBaseVersion`.

Reply:
OK, get it. There is no need to call `DMFilePackFilter::loadFrom` which is kind of complex. How about creating `DMFilePackFilterResult` according to `read_packs_id` directly?

Reply:
ok